File: backend/src/main/java/com/genesis/backend/infrastructure/adapters/in/web/ChatController.java
package com.genesis.backend.infrastructure.adapters.in.web;

import com.genesis.backend.domain.model.Message;
import com.genesis.backend.domain.model.MessageRole;
import com.genesis.backend.domain.model.ToolCall;
import com.genesis.backend.domain.ports.in.ChatUseCasePort;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/api/chat")
@CrossOrigin(origins = "*")
public class ChatController {

    private final ChatUseCasePort chatUseCase;

    public ChatController(ChatUseCasePort chatUseCase) {
        this.chatUseCase = chatUseCase;
    }

    @PostMapping
    public AiResponseDto processChat(@RequestBody ChatRequest request) {
        List<Message> domainMessages = request.messages().stream().map(dto -> {
            MessageRole role;
            try {
                role = MessageRole.valueOf(dto.role().toUpperCase());
            } catch (IllegalArgumentException e) {
                role = MessageRole.USER;
            }
            return new Message(role, dto.content());
        }).toList();

        var aiResponse = chatUseCase.processChat(domainMessages);

        List<ToolCallDto> toolCallDtos = aiResponse.toolCalls().stream().map(tc -> {
            if (tc instanceof ToolCall.UpdateSpecSection uss) {
                return new ToolCallDto("update_spec_section", uss.id(), uss.sectionId(), uss.content(), uss.status(), null);
            } else if (tc instanceof ToolCall.UpdateProjectName upn) {
                return new ToolCallDto("update_project_name", upn.id(), null, null, null, upn.name());
            }
            return null;
        }).toList();

        return new AiResponseDto(aiResponse.textContent(), toolCallDtos);
    }
}

File: backend/src/main/kotlin/com/genesis/backend/infrastructure/adapters/in/web/ChatController.kt
package com.genesis.backend.infrastructure.adapters.`in`.web

import com.genesis.backend.domain.model.Message
import com.genesis.backend.domain.ports.`in`.ChatUseCasePort
import io.ktor.server.application.*
import io.ktor.server.request.*
import io.ktor.server.response.*
import io.ktor.server.routing.*
import kotlinx.serialization.Serializable

@Serializable
data class ChatRequest(val messages: List<MessageDto>)

@Serializable
data class MessageDto(val role: String, val content: String)

@Serializable
data class AiResponseDto(
    val textContent: String?,
    val toolCalls: List<ToolCallDto>
)

@Serializable
data class ToolCallDto(
    val type: String,
    val id: String,
    val sectionId: String? = null,
    val content: String? = null,
    val status: String? = null,
    val name: String? = null
)

fun Route.chatRoutes(chatUseCase: ChatUseCasePort) {
    post("/api/chat") {
        val request = call.receive<ChatRequest>()
        
        val domainMessages = request.messages.map {
            val role = when (it.role.lowercase()) {
                "user" -> com.genesis.backend.domain.model.MessageRole.USER
                "model" -> com.genesis.backend.domain.model.MessageRole.MODEL
                "system" -> com.genesis.backend.domain.model.MessageRole.SYSTEM
                else -> com.genesis.backend.domain.model.MessageRole.USER
            }
            Message(role = role, content = it.content)
        }

        val aiResponse = chatUseCase.processChat(domainMessages)
        
        val responseDto = AiResponseDto(
            textContent = aiResponse.textContent,
            toolCalls = aiResponse.toolCalls.map { toolCall ->
                when (toolCall) {
                    is com.genesis.backend.domain.model.ToolCall.UpdateSpecSection -> {
                        ToolCallDto(
                            type = "update_spec_section",
                            id = toolCall.id,
                            sectionId = toolCall.sectionId,
                            content = toolCall.content,
                            status = toolCall.status
                        )
                    }
                    is com.genesis.backend.domain.model.ToolCall.UpdateProjectName -> {
                        ToolCallDto(
                            type = "update_project_name",
                            id = toolCall.id,
                            name = toolCall.name
                        )
                    }
                }
            }
        )
        
        call.respond(responseDto)
    }
}
